src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigWatcher.java

```diff
@@ -123,248 +131,352 @@ public void doRun() {
+        if (items != null) {
+            for (BuildConfig buildConfig : items) {
+                try {
+                    upsertJob(buildConfig);
+                } catch (Exception e) {
+                    logger.log(SEVERE, "Failed to update job", e);
+                }
+            }
+        }
+    }
+
+    @SuppressFBWarnings("SF_SWITCH_NO_DEFAULT")
+    @Override
+    public synchronized void eventReceived(Watcher.Action action,
+            BuildConfig buildConfig) {
+        try {
+            switch (action) {
+            case ADDED:
+                upsertJob(buildConfig);
+                break;
+            case DELETED:
+                deleteEventToJenkinsJob(buildConfig);
+                break;
+            case MODIFIED:
+                modifyEventToJenkinsJob(buildConfig);
+                break;
+            }
+            // if bc event came after build events, let's
+            // poke the BuildWatcher builds with no BC list to create job
+            // runs
+            BuildWatcher.flushBuildsWithNoBCList();
+            // now, if the build event was lost and never received, builds
+            // will stay in
+            // new for 5 minutes ... let's launch a background thread to
+            // clean them up
+            // at a quicker interval than the default 5 minute general build
+            // relist function
+            if (action == Watcher.Action.ADDED) {
+                Runnable backupBuildQuery = new SafeTimerTask() {
+                    @Override
+                    public void doRun() {
+                        if (!CredentialsUtils.hasCredentials()) {
+                            logger.fine("No Openshift Token credential defined.");
+                            return;
+                        }
+                        BuildList buildList = getAuthenticatedOpenShiftClient()
+                                .builds()
+                                .inNamespace(
+                                        buildConfig.getMetadata()
+                                                .getNamespace())
+                                .withField(OPENSHIFT_BUILD_STATUS_FIELD,
+                                        BuildPhases.NEW)
+                                .withLabel(OPENSHIFT_LABELS_BUILD_CONFIG_NAME,
+                                        buildConfig.getMetadata().getName())
+                                .list();
+                        if (buildList.getItems().size() > 0) {
+                            logger.info("build backup query for "
+                                    + buildConfig.getMetadata().getName()
+                                    + " found new builds");
+                            BuildWatcher.onInitialBuilds(buildList);
+                        }
+                    }
+                };
+                Timer.get().schedule(backupBuildQuery, 10 * 1000,
+                        TimeUnit.MILLISECONDS);
```

Comment:
Is there any risk of a race condition here if the build event comes in later?
ie:
1) buildconfig created
2) build created
3) this task is scheduled
4) build event shows up
5) race between 3+4 with potentially both launching corresponding jenkins jobs

or do we have sufficient locking around the creation of jenkins jobs (plus re-validating the build is still in a new state)?

Reply:
Yep, good question.  Fortunately, I believe we have a good answer.

We have sufficient locking protection.  Ultimately it is no different then how the relist and event paths already work today.

Note this code:  https://github.com/openshift/jenkins-sync-plugin/blob/master/src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java#L301-L313

With the static method `addEventToJenkinsJobRun` synchronized, that is protection point 1.  It is single threaded for sure.  

Then, there is the check where if the build is no longer new, simply return.  As part of launching the jenkins job run, the build gets moved out of new state.

There is further protection later on at https://github.com/openshift/jenkins-sync-plugin/blob/master/src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java#L287-L300

If the job run corresponding to the build is already active, we punt there as well.

Reply:
ok cool.